lib/recipe/bloc/recipe_bloc.dart

```diff
@@ -70,22 +76,121 @@ class RecipeBloc extends Bloc<RecipeEvent, RecipeState> {
+    measurements.add(measurement);
+
+    Recipe recipe =
+        recipes[changedRecipeIndex].copyWith(measurements: measurements);
 
+    recipes.removeAt(changedRecipeIndex);
+    recipes.insert(changedRecipeIndex, recipe);
+
+    emit(state.copyWith(recipes: recipes));
+  }
+
+  void _onRecipeMeasurementDeleted(
+      RecipeMeasurementDeleted event, Emitter<RecipeState> emit) {
+    List<Recipe> recipes = List.from(state.recipes);
     int changedRecipeIndex = recipes.indexOf(event.recipe);
 
-    Recipe recipe = recipes[changedRecipeIndex].copyWith(
-        recipeVolume: RecipeVolume.dirty(event.recipeVolume), saved: false);
+    List<Measurement> measurements =
+        List.from(recipes[changedRecipeIndex].measurements);
+    int changedmeasurementIndex = event.measurementIndex;
+
+    if (measurements.length > 1) {
+      measurements.removeAt(changedmeasurementIndex);
+    }
+
+    Recipe recipe =
+        recipes[changedRecipeIndex].copyWith(measurements: measurements);
+
+    recipes.removeAt(changedRecipeIndex);
+    recipes.insert(changedRecipeIndex, recipe);
+
+    emit(state.copyWith(recipes: recipes));
+  }
+
+  void _onRecipeMeasurementMethodChanged(
+      RecipeMeasurementMethodChanged event, Emitter<RecipeState> emit) {
+    List<Recipe> recipes = List.from(state.recipes);
+    int changedRecipeIndex = recipes.indexOf(event.recipe);
+
+    List<Measurement> measurements =
+        List.from(recipes[changedRecipeIndex].measurements);
+    int changedmeasurementIndex = event.measurementIndex;
+
+    Measurement measurement = measurements[changedmeasurementIndex]
+        .copyWith(measurementMethod: event.measurementMethod);
+
+    measurements.removeAt(changedmeasurementIndex);
+    measurements.insert(changedmeasurementIndex, measurement);
+
+    Recipe recipe =
+        recipes[changedRecipeIndex].copyWith(measurements: measurements);
+
+    recipes.removeAt(changedRecipeIndex);
+    recipes.insert(changedRecipeIndex, recipe);
+
+    emit(state.copyWith(recipes: recipes));
+  }
+
+  void _onRecipeMeasurementUnitChanged(
+      RecipeMeasurementUnitChanged event, Emitter<RecipeState> emit) {
+    List<Recipe> recipes = List.from(state.recipes);
+    int changedRecipeIndex = recipes.indexOf(event.recipe);
+
+    List<Measurement> measurements =
+        List.from(recipes[changedRecipeIndex].measurements);
+    int changedmeasurementIndex = event.measurementIndex;
+
+    Measurement measurement = measurements[changedmeasurementIndex]
+        .copyWith(measurementUnit: event.measurementUnit);
+
+    measurements.removeAt(changedmeasurementIndex);
+    measurements.insert(changedmeasurementIndex, measurement);
+
+    Recipe recipe =
+        recipes[changedRecipeIndex].copyWith(measurements: measurements);
+
+    recipes.removeAt(changedRecipeIndex);
+    recipes.insert(changedRecipeIndex, recipe);
+
+    emit(state.copyWith(recipes: recipes));
+  }
+
+  void _onRecipeMeasurementVolumeChanged(
```

Comment:
As I'll elaborate in later comment, I think it would be more true to its name to have `MeasurementValue` instead of `MeasurementVolume`